File: mobile/src/main/java/com/backdoor/moove/core/fragments/NotificationSettingsFragment.java
package com.backdoor.moove.core.fragments;


import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.backdoor.moove.R;
import com.backdoor.moove.core.consts.Configs;
import com.backdoor.moove.core.consts.Constants;
import com.backdoor.moove.core.consts.Prefs;
import com.backdoor.moove.core.helper.Dialogues;
import com.backdoor.moove.core.helper.Module;
import com.backdoor.moove.core.helper.Permissions;
import com.backdoor.moove.core.helper.SharedPrefs;
import com.backdoor.moove.core.views.PrefsView;

import java.io.File;

public class NotificationSettingsFragment extends Fragment implements View.OnClickListener,
        DialogInterface.OnDismissListener {

    private SharedPrefs sPrefs;
    private ActionBar ab;
    private TextView locale;

    private PrefsView blurPrefs, vibrationOptionPrefs, infiniteVibrateOptionPrefs,
            soundOptionPrefs, infiniteSoundOptionPrefs, ttsPrefs, wakeScreenOptionPrefs,
            unlockScreenPrefs, silentSMSOptionPrefs, ledPrefs, chooseSoundPrefs,
            chooseLedColorPrefs, silentCallOptionPrefs;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.settings_notification, container, false);
        ab = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (ab != null) {
            ab.setTitle(R.string.notification);
        }
        sPrefs = SharedPrefs.getInstance(getActivity());
        if (Module.isLollipop()) {
            rootView.findViewById(R.id.imageCard).setElevation(Configs.CARD_ELEVATION);
            rootView.findViewById(R.id.soundCard).setElevation(Configs.CARD_ELEVATION);
            rootView.findViewById(R.id.systemCard).setElevation(Configs.CARD_ELEVATION);
        }
        TextView selectImage = rootView.findViewById(R.id.selectImage);
        selectImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Dialogues.imageDialog(getActivity(), new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialogInterface) {

                    }
                });
            }
        });

        blurPrefs = rootView.findViewById(R.id.blurPrefs);
        blurPrefs.setOnClickListener(this);
        blurPrefs.setChecked(sPrefs.loadBoolean(Prefs.REMINDER_IMAGE_BLUR));
        blurPrefs.setEnabled(true);

        vibrationOptionPrefs = rootView.findViewById(R.id.vibrationOptionPrefs);
        vibrationOptionPrefs.setOnClickListener(this);
        vibrationOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.VIBRATION_STATUS));

        infiniteVibrateOptionPrefs = rootView.findViewById(R.id.infiniteVibrateOptionPrefs);
        infiniteVibrateOptionPrefs.setOnClickListener(this);
        infiniteVibrateOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.INFINITE_VIBRATION));

        soundOptionPrefs = rootView.findViewById(R.id.soundOptionPrefs);
        soundOptionPrefs.setOnClickListener(this);
        soundOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.SILENT_SOUND));

        infiniteSoundOptionPrefs = rootView.findViewById(R.id.infiniteSoundOptionPrefs);
        infiniteSoundOptionPrefs.setOnClickListener(this);
        infiniteSoundOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.INFINITE_SOUND));

        ttsPrefs = rootView.findViewById(R.id.ttsPrefs);
        ttsPrefs.setOnClickListener(this);
        ttsPrefs.setChecked(sPrefs.loadBoolean(Prefs.TTS));

        wakeScreenOptionPrefs = rootView.findViewById(R.id.wakeScreenOptionPrefs);
        wakeScreenOptionPrefs.setOnClickListener(this);
        wakeScreenOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.WAKE_STATUS));

        unlockScreenPrefs = rootView.findViewById(R.id.unlockScreenPrefs);
        unlockScreenPrefs.setOnClickListener(this);
        unlockScreenPrefs.setChecked(sPrefs.loadBoolean(Prefs.UNLOCK_DEVICE));

        silentSMSOptionPrefs = rootView.findViewById(R.id.silentSMSOptionPrefs);
        silentSMSOptionPrefs.setOnClickListener(this);
        silentSMSOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.SILENT_SMS));

        silentCallOptionPrefs = rootView.findViewById(R.id.silentCallOptionPrefs);
        silentCallOptionPrefs.setOnClickListener(this);
        silentCallOptionPrefs.setChecked(sPrefs.loadBoolean(Prefs.SILENT_CALL));

        chooseSoundPrefs = rootView.findViewById(R.id.chooseSoundPrefs);
        chooseSoundPrefs.setOnClickListener(this);

        showMelody();

        TextView volume = rootView.findViewById(R.id.volume);
        volume.setOnClickListener(this);

        locale = rootView.findViewById(R.id.locale);
        locale.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Dialogues.ttsLocale(getActivity(), Prefs.TTS_LOCALE);
            }
        });

        ledPrefs = rootView.findViewById(R.id.ledPrefs);
        chooseLedColorPrefs = rootView.findViewById(R.id.chooseLedColorPrefs);

        checkVibrate();

        checkTTS();

        ledPrefs.setOnClickListener(this);
        ledPrefs.setVisibility(View.VISIBLE);
        ledPrefs.setChecked(sPrefs.loadBoolean(Prefs.LED_STATUS));

        chooseLedColorPrefs.setVisibility(View.VISIBLE);
        chooseLedColorPrefs.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Dialogues.ledColor(getActivity());
            }
        });

        checkEnabling();

        return rootView;
    }

    private void checkTTS() {
        if (ttsPrefs.isChecked()) {
            locale.setEnabled(true);
        } else {
            locale.setEnabled(false);
        }
    }

    private void checkVibrate() {
        if (vibrationOptionPrefs.isChecked()) {
            infiniteVibrateOptionPrefs.setEnabled(true);
        } else {
            infiniteVibrateOptionPrefs.setEnabled(false);
        }
    }

    private void showMelody() {
        if (sPrefs.loadBoolean(Prefs.CUSTOM_SOUND)) {
            if (sPrefs.isString(Prefs.CUSTOM_SOUND_FILE)) {
                String path = sPrefs.loadPrefs(Prefs.CUSTOM_SOUND_FILE);
                if (!path.matches("")) {
                    File sound = new File(path);
                    if (sound.exists()) {
                        String fileName = sound.getName();
                        int pos = fileName.lastIndexOf(".");
                        String fileNameS = fileName.substring(0, pos);
                        chooseSoundPrefs.setDetailText(fileNameS);
                    } else {
                        chooseSoundPrefs.setDetailText(getActivity().getString(R.string.default_string));
                    }
                } else {
                    chooseSoundPrefs.setDetailText(getActivity().getString(R.string.default_string));
                }
            }
        } else {
            chooseSoundPrefs.setDetailText(getResources().getString(R.string.default_string));
        }
    }

    private void vibrationChange() {
        if (vibrationOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.VIBRATION_STATUS, false);
            vibrationOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.VIBRATION_STATUS, true);
            vibrationOptionPrefs.setChecked(true);
        }
        checkVibrate();
    }

    private void infiniteVibrationChange() {
        if (infiniteVibrateOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.INFINITE_VIBRATION, false);
            infiniteVibrateOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.INFINITE_VIBRATION, true);
            infiniteVibrateOptionPrefs.setChecked(true);
        }
    }

    private void ttsChange() {
        if (ttsPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.TTS, false);
            ttsPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.TTS, true);
            ttsPrefs.setChecked(true);
            Dialogues.ttsLocale(getActivity(), Prefs.TTS_LOCALE);
        }
        checkTTS();
    }

    private void blurChange() {
        if (blurPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.REMINDER_IMAGE_BLUR, false);
            blurPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.REMINDER_IMAGE_BLUR, true);
            blurPrefs.setChecked(true);
        }
    }

    private void soundChange() {
        if (soundOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.SILENT_SOUND, false);
            soundOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.SILENT_SOUND, true);
            soundOptionPrefs.setChecked(true);
        }
    }

    private void infiniteSoundChange() {
        if (infiniteSoundOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.INFINITE_SOUND, false);
            infiniteSoundOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.INFINITE_SOUND, true);
            infiniteSoundOptionPrefs.setChecked(true);
        }
    }

    private void checkEnabling() {
        if (ledPrefs.isChecked()) {
            chooseLedColorPrefs.setEnabled(true);
        } else {
            chooseLedColorPrefs.setEnabled(false);
        }
    }

    private void ledChange() {
        if (ledPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.LED_STATUS, false);
            ledPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.LED_STATUS, true);
            ledPrefs.setChecked(true);
        }
        checkEnabling();
    }

    private void silentSMSChange() {
        if (silentSMSOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.SILENT_SMS, false);
            silentSMSOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.SILENT_SMS, true);
            silentSMSOptionPrefs.setChecked(true);
        }
    }

    private void silentCallChange() {
        if (silentCallOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.SILENT_CALL, false);
            silentCallOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.SILENT_CALL, true);
            silentCallOptionPrefs.setChecked(true);
        }
    }

    private void wakeChange() {
        if (wakeScreenOptionPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.WAKE_STATUS, false);
            wakeScreenOptionPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.WAKE_STATUS, true);
            wakeScreenOptionPrefs.setChecked(true);
        }
    }

    private void unlockChange() {
        if (unlockScreenPrefs.isChecked()) {
            sPrefs.saveBoolean(Prefs.UNLOCK_DEVICE, false);
            unlockScreenPrefs.setChecked(false);
        } else {
            sPrefs.saveBoolean(Prefs.UNLOCK_DEVICE, true);
            unlockScreenPrefs.setChecked(true);
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        ab = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (ab != null) {
            ab.setTitle(R.string.settings);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.blurPrefs:
                blurChange();
                break;
            case R.id.vibrationOptionPrefs:
                vibrationChange();
                break;
            case R.id.soundOptionPrefs:
                soundChange();
                break;
            case R.id.infiniteSoundOptionPrefs:
                infiniteSoundChange();
                break;
            case R.id.chooseSoundPrefs:
                if (Permissions.checkPermission(getActivity(), Permissions.READ_EXTERNAL)) {
                    Dialogues.melodyType(getActivity(), Prefs.CUSTOM_SOUND, 201);
                } else {
                    Permissions.requestPermission(getActivity(), 103,
                            Permissions.READ_EXTERNAL, Permissions.WRITE_EXTERNAL);
                }
                break;
            case R.id.infiniteVibrateOptionPrefs:
                infiniteVibrationChange();
                break;
            case R.id.silentSMSOptionPrefs:
                silentSMSChange();
                break;
            case R.id.wakeScreenOptionPrefs:
                wakeChange();
                break;
            case R.id.unlockScreenPrefs:
                unlockChange();
                break;
            case R.id.volume:
                Dialogues.dialogWithSeek(getActivity(), 25, Prefs.VOLUME, getString(R.string.volume), this);
                break;
            case R.id.ttsPrefs:
                ttsChange();
                break;
            case R.id.ledPrefs:
                ledChange();
                break;
            case R.id.silentCallOptionPrefs:
                silentCallChange();
                break;
        }
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        showMelody();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        showMelody();
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case 101:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
                    if (Module.isKitkat()) {
                        intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
                        intent.addCategory(Intent.CATEGORY_OPENABLE);
                    }
                    intent.setType("image/*");
                    Intent chooser = Intent.createChooser(intent, getActivity().getString(R.string.select_image));
                    getActivity().startActivityForResult(chooser, Constants.ACTION_REQUEST_GALLERY);
                } else {
                    Permissions.showInfo(getActivity(), Permissions.READ_CALENDAR);
                }
                break;
        }
    }
}

File: mobile/src/main/java/com/backdoor/moove/core/fragments/NotificationSettingsFragment.kt
package com.backdoor.moove.core.fragments


import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.backdoor.moove.R
import com.backdoor.moove.core.consts.Configs
import com.backdoor.moove.core.consts.Constants
import com.backdoor.moove.core.consts.Prefs
import com.backdoor.moove.core.helper.Dialogues
import com.backdoor.moove.core.helper.Module
import com.backdoor.moove.core.helper.Permissions
import com.backdoor.moove.core.helper.SharedPrefs
import com.backdoor.moove.core.views.PrefsView

import java.io.File

class NotificationSettingsFragment : Fragment(), View.OnClickListener, DialogInterface.OnDismissListener {

    private var sPrefs: SharedPrefs? = null
    private var ab: ActionBar? = null
    private var locale: TextView? = null

    private var blurPrefs: PrefsView? = null
    private var vibrationOptionPrefs: PrefsView? = null
    private var infiniteVibrateOptionPrefs: PrefsView? = null
    private var soundOptionPrefs: PrefsView? = null
    private var infiniteSoundOptionPrefs: PrefsView? = null
    private var ttsPrefs: PrefsView? = null
    private var wakeScreenOptionPrefs: PrefsView? = null
    private var unlockScreenPrefs: PrefsView? = null
    private var silentSMSOptionPrefs: PrefsView? = null
    private var ledPrefs: PrefsView? = null
    private var chooseSoundPrefs: PrefsView? = null
    private var chooseLedColorPrefs: PrefsView? = null
    private var silentCallOptionPrefs: PrefsView? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.settings_notification, container, false)
        ab = (activity as AppCompatActivity).supportActionBar
        if (ab != null) {
            ab!!.setTitle(R.string.notification)
        }
        sPrefs = SharedPrefs.getInstance(activity)
        if (Module.isLollipop) {
            rootView.findViewById<View>(R.id.imageCard).elevation = Configs.CARD_ELEVATION
            rootView.findViewById<View>(R.id.soundCard).elevation = Configs.CARD_ELEVATION
            rootView.findViewById<View>(R.id.systemCard).elevation = Configs.CARD_ELEVATION
        }
        val selectImage = rootView.findViewById<TextView>(R.id.selectImage)
        selectImage.setOnClickListener { v ->
            Dialogues.imageDialog(activity) { dialogInterface ->

            }
        }

        blurPrefs = rootView.findViewById(R.id.blurPrefs)
        blurPrefs!!.setOnClickListener(this)
        blurPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.REMINDER_IMAGE_BLUR)
        blurPrefs!!.isEnabled = true

        vibrationOptionPrefs = rootView.findViewById(R.id.vibrationOptionPrefs)
        vibrationOptionPrefs!!.setOnClickListener(this)
        vibrationOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.VIBRATION_STATUS)

        infiniteVibrateOptionPrefs = rootView.findViewById(R.id.infiniteVibrateOptionPrefs)
        infiniteVibrateOptionPrefs!!.setOnClickListener(this)
        infiniteVibrateOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.INFINITE_VIBRATION)

        soundOptionPrefs = rootView.findViewById(R.id.soundOptionPrefs)
        soundOptionPrefs!!.setOnClickListener(this)
        soundOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.SILENT_SOUND)

        infiniteSoundOptionPrefs = rootView.findViewById(R.id.infiniteSoundOptionPrefs)
        infiniteSoundOptionPrefs!!.setOnClickListener(this)
        infiniteSoundOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.INFINITE_SOUND)

        ttsPrefs = rootView.findViewById(R.id.ttsPrefs)
        ttsPrefs!!.setOnClickListener(this)
        ttsPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.TTS)

        wakeScreenOptionPrefs = rootView.findViewById(R.id.wakeScreenOptionPrefs)
        wakeScreenOptionPrefs!!.setOnClickListener(this)
        wakeScreenOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.WAKE_STATUS)

        unlockScreenPrefs = rootView.findViewById(R.id.unlockScreenPrefs)
        unlockScreenPrefs!!.setOnClickListener(this)
        unlockScreenPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.UNLOCK_DEVICE)

        silentSMSOptionPrefs = rootView.findViewById(R.id.silentSMSOptionPrefs)
        silentSMSOptionPrefs!!.setOnClickListener(this)
        silentSMSOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.SILENT_SMS)

        silentCallOptionPrefs = rootView.findViewById(R.id.silentCallOptionPrefs)
        silentCallOptionPrefs!!.setOnClickListener(this)
        silentCallOptionPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.SILENT_CALL)

        chooseSoundPrefs = rootView.findViewById(R.id.chooseSoundPrefs)
        chooseSoundPrefs!!.setOnClickListener(this)

        showMelody()

        val volume = rootView.findViewById<TextView>(R.id.volume)
        volume.setOnClickListener(this)

        locale = rootView.findViewById(R.id.locale)
        locale!!.setOnClickListener { v -> Dialogues.ttsLocale(activity, Prefs.TTS_LOCALE) }

        ledPrefs = rootView.findViewById(R.id.ledPrefs)
        chooseLedColorPrefs = rootView.findViewById(R.id.chooseLedColorPrefs)

        checkVibrate()

        checkTTS()

        ledPrefs!!.setOnClickListener(this)
        ledPrefs!!.visibility = View.VISIBLE
        ledPrefs!!.isChecked = sPrefs!!.loadBoolean(Prefs.LED_STATUS)

        chooseLedColorPrefs!!.visibility = View.VISIBLE
        chooseLedColorPrefs!!.setOnClickListener { view -> Dialogues.ledColor(activity) }

        checkEnabling()

        return rootView
    }

    private fun checkTTS() {
        if (ttsPrefs!!.isChecked) {
            locale!!.isEnabled = true
        } else {
            locale!!.isEnabled = false
        }
    }

    private fun checkVibrate() {
        if (vibrationOptionPrefs!!.isChecked) {
            infiniteVibrateOptionPrefs!!.isEnabled = true
        } else {
            infiniteVibrateOptionPrefs!!.isEnabled = false
        }
    }

    private fun showMelody() {
        if (sPrefs!!.loadBoolean(Prefs.CUSTOM_SOUND)) {
            if (sPrefs!!.isString(Prefs.CUSTOM_SOUND_FILE)) {
                val path = sPrefs!!.loadPrefs(Prefs.CUSTOM_SOUND_FILE)
                if (!path!!.matches("".toRegex())) {
                    val sound = File(path)
                    if (sound.exists()) {
                        val fileName = sound.name
                        val pos = fileName.lastIndexOf(".")
                        val fileNameS = fileName.substring(0, pos)
                        chooseSoundPrefs!!.setDetailText(fileNameS)
                    } else {
                        chooseSoundPrefs!!.setDetailText(activity!!.getString(R.string.default_string))
                    }
                } else {
                    chooseSoundPrefs!!.setDetailText(activity!!.getString(R.string.default_string))
                }
            }
        } else {
            chooseSoundPrefs!!.setDetailText(resources.getString(R.string.default_string))
        }
    }

    private fun vibrationChange() {
        if (vibrationOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.VIBRATION_STATUS, false)
            vibrationOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.VIBRATION_STATUS, true)
            vibrationOptionPrefs!!.isChecked = true
        }
        checkVibrate()
    }

    private fun infiniteVibrationChange() {
        if (infiniteVibrateOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.INFINITE_VIBRATION, false)
            infiniteVibrateOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.INFINITE_VIBRATION, true)
            infiniteVibrateOptionPrefs!!.isChecked = true
        }
    }

    private fun ttsChange() {
        if (ttsPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.TTS, false)
            ttsPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.TTS, true)
            ttsPrefs!!.isChecked = true
            Dialogues.ttsLocale(activity, Prefs.TTS_LOCALE)
        }
        checkTTS()
    }

    private fun blurChange() {
        if (blurPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.REMINDER_IMAGE_BLUR, false)
            blurPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.REMINDER_IMAGE_BLUR, true)
            blurPrefs!!.isChecked = true
        }
    }

    private fun soundChange() {
        if (soundOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.SILENT_SOUND, false)
            soundOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.SILENT_SOUND, true)
            soundOptionPrefs!!.isChecked = true
        }
    }

    private fun infiniteSoundChange() {
        if (infiniteSoundOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.INFINITE_SOUND, false)
            infiniteSoundOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.INFINITE_SOUND, true)
            infiniteSoundOptionPrefs!!.isChecked = true
        }
    }

    private fun checkEnabling() {
        if (ledPrefs!!.isChecked) {
            chooseLedColorPrefs!!.isEnabled = true
        } else {
            chooseLedColorPrefs!!.isEnabled = false
        }
    }

    private fun ledChange() {
        if (ledPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.LED_STATUS, false)
            ledPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.LED_STATUS, true)
            ledPrefs!!.isChecked = true
        }
        checkEnabling()
    }

    private fun silentSMSChange() {
        if (silentSMSOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.SILENT_SMS, false)
            silentSMSOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.SILENT_SMS, true)
            silentSMSOptionPrefs!!.isChecked = true
        }
    }

    private fun silentCallChange() {
        if (silentCallOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.SILENT_CALL, false)
            silentCallOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.SILENT_CALL, true)
            silentCallOptionPrefs!!.isChecked = true
        }
    }

    private fun wakeChange() {
        if (wakeScreenOptionPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.WAKE_STATUS, false)
            wakeScreenOptionPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.WAKE_STATUS, true)
            wakeScreenOptionPrefs!!.isChecked = true
        }
    }

    private fun unlockChange() {
        if (unlockScreenPrefs!!.isChecked) {
            sPrefs!!.saveBoolean(Prefs.UNLOCK_DEVICE, false)
            unlockScreenPrefs!!.isChecked = false
        } else {
            sPrefs!!.saveBoolean(Prefs.UNLOCK_DEVICE, true)
            unlockScreenPrefs!!.isChecked = true
        }
    }

    override fun onDetach() {
        super.onDetach()
        ab = (activity as AppCompatActivity).supportActionBar
        if (ab != null) {
            ab!!.setTitle(R.string.settings)
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.blurPrefs -> blurChange()
            R.id.vibrationOptionPrefs -> vibrationChange()
            R.id.soundOptionPrefs -> soundChange()
            R.id.infiniteSoundOptionPrefs -> infiniteSoundChange()
            R.id.chooseSoundPrefs -> if (Permissions.checkPermission(activity, Permissions.READ_EXTERNAL)) {
                Dialogues.melodyType(activity, Prefs.CUSTOM_SOUND, 201)
            } else {
                Permissions.requestPermission(activity, 103,
                        Permissions.READ_EXTERNAL, Permissions.WRITE_EXTERNAL)
            }
            R.id.infiniteVibrateOptionPrefs -> infiniteVibrationChange()
            R.id.silentSMSOptionPrefs -> silentSMSChange()
            R.id.wakeScreenOptionPrefs -> wakeChange()
            R.id.unlockScreenPrefs -> unlockChange()
            R.id.volume -> Dialogues.dialogWithSeek(activity, 25, Prefs.VOLUME, getString(R.string.volume), this)
            R.id.ttsPrefs -> ttsChange()
            R.id.ledPrefs -> ledChange()
            R.id.silentCallOptionPrefs -> silentCallChange()
        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        showMelody()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        showMelody()
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (grantResults.size == 0) return
        when (requestCode) {
            101 -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                var intent = Intent(Intent.ACTION_GET_CONTENT)
                if (Module.isKitkat) {
                    intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
                    intent.addCategory(Intent.CATEGORY_OPENABLE)
                }
                intent.type = "image/*"
                val chooser = Intent.createChooser(intent, activity!!.getString(R.string.select_image))
                activity!!.startActivityForResult(chooser, Constants.ACTION_REQUEST_GALLERY)
            } else {
                Permissions.showInfo(activity, Permissions.READ_CALENDAR)
            }
        }
    }
}
